procrastinate/utils.py

```diff
@@ -307,3 +287,173 @@ async def _inner_coro() -> U:
             return self._return_value
 
         return _inner_coro().__await__()
+
+
+class EndMain(Exception):
+    pass
+
+
+@attr.dataclass()
+class ExceptionRecord:
+    task: asyncio.Task
+    exc: Exception
+
+
+async def run_tasks(
```

Comment:
All in all this function is 150 lines long, but there's a lot of doc/comments. Also, I'm not sure how to split it more efficiently. The base logic would be shorter but it order to log in all the strategic places, and because asyncio.Task supports task names only on 3.8+, we end up with more complex code.